README.md
# projeto-programmatic-generated-images-library

## Descrição

O trabalho consiste na implementação de uma biblioteca que permitirá a geração de imagens de forma programática através da utilização de operações e criações de estruturas baseadas em uma imagem geradora inicial, como visualizado no trabalho 'Functional Geometry' feito por Peter Henderson (https://eprints.soton.ac.uk/257577/1/funcgeo2.pdf). 

O objetivo é que a biblioteca permita tanto a geração de imagens a partir de uma imagem base que atenda certos padrões para que o resultado tenha as características esperadas, quanto a possível adapatação e/ou geração de novas imagens que inicialmente não seriam boas canditadas para esse tipo de algoritmo, padrões esses exemplificados também no trabalho 'Functional Geometry'.

ChangeLog.md
# Changelog for projeto-programmatic-generated-images-library

## Unreleased changes


test/Spec.hs
import Draw (saveDrawing)
import TestDrawings
import CreateDrawings (white, colorsD)
import Models

test_img = saveDrawing "img_tests/img.png" 1000 white img_base
test_rot = saveDrawing "img_tests/rot.png" 1000 white img_rot
test_rot45 = saveDrawing "img_tests/rot45.png" 1000 white img_rot45
test_img2 = saveDrawing "img_tests/img2.png" 1000 white img_img2
test_img3 = saveDrawing "img_tests/img3.png" 1000 white img_img3
test_u = saveDrawing "img_tests/u.png" 1000 white img_u
test_t = saveDrawing "img_tests/t.png" 1000 white img_t
test_v = saveDrawing "img_tests/v.png" 1000 white img_v
test_quartet = saveDrawing "img_tests/quartet.png" 1000 white img_quartet
test_side2 = saveDrawing "img_tests/side2.png" 1000 white img_side2
test_corner2 = saveDrawing "img_tests/corner2.png" 1000 white img_corner2
test_squarelimit2 = saveDrawing "img_tests/squarelimit2.png" 1000 white img_squarelimit2
test_squarelimit3 = saveDrawing "img_tests/squarelimit3.png" 1000 white img_squarelimit3
test_squarelimit4 = saveDrawing "img_tests/squarelimit4.png" 1000 white img_squarelimit4

test_simple_img = genArcs "img_tests/simplefig.png" 0 seedE

test_simple_img2 = genArcsNew testArc1 "img_tests/simplefig2.png" 0 (seedE+10) colorsD

test_simple_img3 = genArcsNew testArc "img_tests/simplefig3.png" 0 (seedE+4) colorsE
  where testArc = createNewArc 0.6 0.2 0.8 0.4

test_squarelimitN = genSquareLimitNew testSquareLimit "img_tests/squarelimitN.png" 3

test_semiCirclesSquaresGen = genSemiCirclesSquares "img_tests/semiCiclesSquaresGen.png" 2 (seedE+120)

test_semiCicles = saveDrawing "img_tests/semiCicles.png" 1000 white semiCicles
test_semiCiclesFill = saveDrawing "img_tests/semiCiclesFill.png" 1000 white semiCiclesFill
test_semiCiclesSquares = saveDrawing "img_tests/semiCiclesSquares.png" 1000 white semiCiclesSquares
test_semiCiclesSquaresFill = saveDrawing "img_tests/semiCiclesSquaresFill.png" 1000 white semiCiclesSquaresFill


main :: IO ()
main = do
  --test_img
  --test_rot
  --test_rot45
  --test_img2
  --test_img3
  --test_u
  --test_t
  test_v
  --test_quartet
  --test_side2
  --test_corner2
  --test_squarelimit2
  --test_squarelimit3
  test_squarelimit4

  test_simple_img
  test_simple_img2
  test_simple_img3

  test_squarelimitN

  test_semiCirclesSquaresGen

  test_semiCicles
  test_semiCiclesFill
  test_semiCiclesSquares
  test_semiCiclesSquaresFill



test/TestDrawings.hs
module TestDrawings where

import Lib
import CreateDrawings
import Shapes
import Draw
import Graphics.Rasterific
import Graphics.Rasterific.Texture

import Codec.Picture( PixelRGBA8( .. ), writePng, Image) -- Tirar

testOut1 = [CubicBezier (V2 1.00 1.00) (V2 0.85 0.95) (V2 0.70 0.95) (V2 0.50 0.98)
           ,CubicBezier (V2 0.50 0.98) (V2 0.30 0.70) (V2 0.15 0.60) (V2 0.00 1.00)
           ]
testOut2 = over (rot45 (Lib.flip (rot testOut1))) (rot (rot (rot45 (Lib.flip (rot testOut1))))) 
testOut3 = transform (multFst (-1)) ((rot.rot.rot) (Lib.flip testOut1))
testSquareLimit = over (over testOut1 testOut2) testOut3

testFig          = fish
img_base         = [withTexture (uniformTexture black) $ do mconcat $ drawLines 1000 2 $  rot testFig]
img_rot          = [withTexture (uniformTexture black) $ do mconcat $ drawLines 1000 2 $  rot testFig]
img_rot45        = [withTexture (uniformTexture black) $ do mconcat $ drawLines 1000 2 $  rot45 testFig]
img_img2         = [withTexture (uniformTexture black) $ do mconcat $ drawLines 1000 2 $  img2 testFig]
img_img3         = [withTexture (uniformTexture black) $ do mconcat $ drawLines 1000 2 $  img3 testFig]
img_u            = [withTexture (uniformTexture black) $ do mconcat $ drawLines 1000 2 $  u testFig]
img_t            = [withTexture (uniformTexture black) $ do mconcat $ drawLines 1000 2 $  t testFig]
img_v            = [withTexture (uniformTexture black) $ do mconcat $ drawLines 1000 2 $  v testFig]
img_quartet      = [withTexture (uniformTexture black) $ do mconcat $ drawLines 1000 2 $  quartet testFig testFig testFig testFig]
img_side2        = [withTexture (uniformTexture black) $ do mconcat $ drawLines 1000 2 $  side 2 testFig]
img_corner2      = [withTexture (uniformTexture black) $ do mconcat $ drawLines 1000 2 $  corner 2 testFig]
img_squarelimit2 = [withTexture (uniformTexture black) $ do mconcat $ drawLines 1000 2 $  squarelimit 2 testFig]
img_squarelimit3 = [withTexture (uniformTexture black) $ do mconcat $ drawLines 1000 2 $  squarelimit 3 testFig]
img_squarelimit4 = [withTexture (uniformTexture black) $ do mconcat $ drawLines 1000 2 $  squarelimit 4 testFig]

testArc1 = [Line (V2 0.5 0.0) (V2 1.0 0.5)]

seedE :: Int
seedE = -2554405803717694884

colorsE :: [PixelRGBA8]
colorsE = [PixelRGBA8 0xE9 0xE3 0xCE 255, PixelRGBA8 0xFF 0x53 0x73 255, 
          PixelRGBA8 0xEE 0xAD 0x2D 255, PixelRGBA8 0x41 0x69 0xE1 255,
          PixelRGBA8 0x5F 0x4B 0x8B 255, PixelRGBA8 0x4B 0x8B 0x3B 255]

semiCiclesFill = [withTexture (uniformTexture fordwine) $ do fill $ scale 1000.0 (makeQuartets 2 (randomArcs seedE) semicircle)]
semiCicles     = [withTexture (uniformTexture black) $ do mconcat $ drawLines 1000 2 $ makeQuartets 2 (randomArcs seedE) semicircle]

semiCiclesSquares     = [withTexture (uniformTexture fordwine) $ do fill $ scale 1000.0 (makeQuartets 2 (randomArcs seedE) semicircle),
  withTexture (uniformTexture black) $ do mconcat $ drawLines 1000 2 $ (makeQuartets 2 (randomArcs seedE) square)]

semiCiclesSquaresFill = [withTexture (uniformTexture fordwine) $ do fill $ scale 1000.0 (makeQuartets 1 (randomArcs seedE) semicircle),
  withTexture (uniformTexture black) $ do mconcat $ drawLines 1000 2 $ (makeQuartets 1 (randomArcs seedE) square)]                     


src/Lib.hs
module Lib where

import Prelude hiding (flip, cycle)
import Graphics.Rasterific
import Graphics.Rasterific.Texture

someFunc :: IO ()
someFunc = putStrLn "someFunc"


-- Tile - bloco unitário

scale :: Transformable a => Float -> a -> a
scale s = transform (fmap (* s))

-- Points auxiliary functions
addSnd :: Float -> Point -> Point
addSnd a p = p + V2 0 a

addFst :: Float -> Point -> Point
addFst a p = p + V2 a 0

multFst :: Float -> Point -> Point
multFst a p = p * V2 a 1

multSnd :: Float -> Point -> Point
multSnd a p = p * V2 1 a

swap :: Point -> Point 
swap (V2 x y) = V2 y x

-- Base operations

flip :: Transformable a => a -> a
flip = transform (addFst 1.multFst (-1))

getProp :: Fractional a => a -> a -> a
getProp f1 f2 = f1/(f1+f2)


over :: [a] -> [a] -> [a]
over = (++)


aboveScaled :: Transformable a => Float -> Float -> [a] -> [a] -> [a]
aboveScaled f1 f2 img1 img2 = trans1 `over` trans2
       where trans1 = transform (multSnd (getProp f1 f2)) img1
             trans2 = transform (addSnd (getProp f1 f2).multSnd (getProp f2 f1)) img2


above :: Transformable a => [a] -> [a] -> [a]
above = aboveScaled 1 1


besideScaled :: Transformable a => Float -> Float -> [a] -> [a] -> [a]
besideScaled f1 f2 img1 img2 = trans1 `over` trans2
       where trans1 = transform (multFst (getProp f1 f2)) img1
             trans2 = transform (addFst (getProp f1 f2).multFst (getProp f2 f1)) img2


beside :: Transformable a => [a] -> [a] -> [a]
beside = besideScaled 1 1


getRandomIO :: Float -> Float
getRandomIO n = 3*n 


-- rotate (x, y) by an angle a (counter-clockwise) = (x2, y2), with
-- x2 = dx + (x - dx) * cos(a) - (dy - y) * sin(a)
-- y2 = dy - ((dy - y) * cos(a)) + ((x - dx) * sin(a))


-- rot (x, y) by 90º 
-- dx = dy = 1/2 
-- x2 = 1/2 - (1/2 - y)*1 = y
-- y2 = 1/2 - (x - 1/2) = 1 - x
rot :: Transformable a => a -> a
rot = transform (addSnd (1).multSnd (-1).swap)

rotn :: Transformable a => Integer -> a -> a
rotn 0 i = i 
rotn n i = rot $ rotn (n-1) i


-- rot (x, y) by 45º 
-- dx = dy = 0
-- r = srqt 2
-- x2 = x* r/2 - (-y * r/2) = r/2 * (x+y)
-- y2 = - ( (-y * r/2)  + (x * r/2) ) r/2 * (y-x)
-- the image must scale by a factor of r, therefore
-- x2 = (x+y)/2
-- y2 = (y-x)/2
rot45 :: Transformable a => a -> a
rot45 = transform (\p -> (*0.5) <$> (addFst (sum p).multFst 0 $ p - (swap p)))
 

blank :: [a]
blank = []

--------------------AQUI
-------------------------------------- 
{-|
genArcs :: FilePath -> Integer -> Int -> [PixelRGBA8] -> IO ()
genArcs path n seed colors = drawAndWriteArcs path (arcLimit n arc) n colors seed


genArcsNew :: (Geometry geom, Transformable geom) => [geom] -> FilePath -> Integer -> Int -> [PixelRGBA8] -> IO ()
genArcsNew img path n seed colors = drawAndWriteArcs path (arcLimit n img) n colors seed

createNewArc :: Float -> Float -> Float -> Float -> [CubicBezier]
createNewArc x1 y1 x2 y2 = [CubicBezier (V2 0.5 0.0) (V2 x1 y1) (V2 x2 y2) (V2 1.0 0.5)]

drawAndWriteArcs :: (Geometry geom, Transformable geom) => FilePath -> [geom] -> Integer -> [PixelRGBA8] -> Int -> IO ()
drawAndWriteArcs path base_img n colors seed = do
    let white = PixelRGBA8 255 255 255 255
        black = PixelRGBA8 0 0 0 255
        img = renderDrawing 1200 1200 white $
            withTexture (uniformTexture black) $ do
                sequence_ (applyFuncs (zip (colorBlocks n colors seed) (blocks n)))
                mconcat $ (\b -> stroke (35/ (2 ** fromInteger (n))) JoinRound (CapRound, CapRound) b) <$> scale 1200 base_img
    writePng path img
-}
----------------------------


src/Draw.hs
module Draw where

import Lib
import Codec.Picture( PixelRGBA8( .. ), writePng, Image)
import Graphics.Rasterific

base_scale = 1000


saveDrawing :: FilePath -> Int -> PixelRGBA8 -> [Drawing PixelRGBA8 ()] -> IO ()
saveDrawing path imgscale backgroundColor drawings = writePng path $ createDrawing imgscale backgroundColor drawings


createDrawing :: Int -> PixelRGBA8 -> [Drawing PixelRGBA8 ()] -> Image PixelRGBA8
createDrawing imgscale backgroundColor drawings = renderDrawing imgscale imgscale backgroundColor $ mconcat drawings


drawLines :: (Geometry geom, Transformable geom) => Int ->  Float -> [geom] -> [Drawing PixelRGBA8 ()]
drawLines appscale strokeWidth img = (\b -> stroke strokeWidth JoinRound (CapRound, CapRound) b) <$> scale (fromIntegral appscale) img


fillForm :: (Geometry geom, Transformable geom) => FillMethod -> [geom] -> Drawing PixelRGBA8 ()
fillForm fillMethod img = fillWithMethod fillMethod img

---------------------------------------------


src/Models.hs
module Models where

import Lib
import Draw
import CreateDrawings
import Shapes
import Codec.Picture( PixelRGBA8( .. ), writePng, Image)
import Graphics.Rasterific
import Graphics.Rasterific.Texture

----- Usable funcs Arcs

genArcs :: FilePath -> Integer -> Int -> IO ()
genArcs path n seed = saveDrawing path scaleSimpleFig white (simpleFig n seed colorsD arc)

genArcsNew :: (Geometry geom, Transformable geom) => [geom] -> FilePath -> Integer -> Int -> [PixelRGBA8] -> IO ()
genArcsNew img path n seed colors = saveDrawing path scaleSimpleFig white (simpleFig n seed colors img)

createNewArc :: Float -> Float -> Float -> Float -> [CubicBezier]
createNewArc x1 y1 x2 y2 = [CubicBezier (V2 0.5 0.0) (V2 x1 y1) (V2 x2 y2) (V2 1.0 0.5)]

---- Helpers Arcs

blocksDrawing :: Integer -> [PixelRGBA8] -> Int -> [Drawing PixelRGBA8 ()]
blocksDrawing n colors seed = applyFuncs (zip (colorBlocks n colors seed) (blocks n))

scaleSimpleFig :: Int
scaleSimpleFig = 1200

simpleFig :: (Geometry geom, Transformable geom) => Integer -> Int -> [PixelRGBA8] -> [geom] -> [Drawing PixelRGBA8 ()]
simpleFig n seed colors img = [withTexture (uniformTexture black) $ do mconcat $ blocksDrawing n colors seed,
                           withTexture (uniformTexture black) $ do mconcat $ drawLines scaleSimpleFig (35/ (2 ** fromInteger n)) (arcLimit n img seed)]

----- Usable funcs Fish

genSquareLimit :: FilePath -> Integer -> IO ()
genSquareLimit path n = saveDrawing path 1000 white (squareLimitFig n fish)

genSquareLimitNew :: (Geometry geom, Transformable geom) => [geom] -> FilePath -> Integer -> IO ()
genSquareLimitNew img path n = saveDrawing path 1000 white (squareLimitFig n img)

---- Helper SquareLimit

squareLimitFig :: (Geometry geom, Transformable geom) => Integer -> [geom] -> [Drawing PixelRGBA8 ()]
squareLimitFig n img = [withTexture (uniformTexture black) $ do mconcat $ drawLines 1000 2 $  squarelimit n img]

----- Usable funcs SemiCirclesSquares

genSemiCirclesSquares :: FilePath -> Integer -> Int -> IO ()
genSemiCirclesSquares path n seed = saveDrawing path 1000 white (semiCirclesSquaresFig n seed)

---- Helper SemiCirclesSquares

semiCirclesSquaresFig :: Integer -> Int -> [Drawing PixelRGBA8 ()]
semiCirclesSquaresFig n seed = [withTexture (uniformTexture fordwine) $ do fill $ scale 1000.0 (makeQuartets n (randomArcs seed) semicircle),
                                withTexture (uniformTexture black) $ do mconcat $ drawLines 1000 2 $ (makeQuartets n (randomArcs seed) square)]                     


src/Shapes.hs
module Shapes where


import Codec.Picture( PixelRGBA8( .. ), writePng)
import Graphics.Rasterific
import Graphics.Rasterific.Texture

triangle :: [Line]
triangle = [ Line (V2 0.00 0.00) (V2 1.00 1.00)
           , Line (V2 0.00 0.00) (V2 0.00 1.00)
           , Line (V2 0.00 1.00) (V2 1.00 1.00)
           ]

fish :: [CubicBezier]
fish = [ (CubicBezier (V2 0.00 0.00) (V2 0.08 0.02) (V2 0.22 0.18) (V2 0.29 0.28))
       , (CubicBezier (V2 0.29 0.28) (V2 0.30 0.36) (V2 0.29 0.43) (V2 0.30 0.50))
       , (CubicBezier (V2 0.30 0.50) (V2 0.34 0.60) (V2 0.43 0.68) (V2 0.50 0.74))
       , (CubicBezier (V2 0.50 0.74) (V2 0.58 0.79) (V2 0.66 0.78) (V2 0.76 0.80))
       , (CubicBezier (V2 0.76 0.80) (V2 0.82 0.88) (V2 0.94 0.95) (V2 1.00 1.00))
       , (CubicBezier (V2 1.00 1.00) (V2 0.90 0.97) (V2 0.81 0.96) (V2 0.76 0.95))
       , (CubicBezier (V2 0.76 0.95) (V2 0.69 0.96) (V2 0.62 0.96) (V2 0.55 0.96))
       , (CubicBezier (V2 0.55 0.96) (V2 0.49 0.90) (V2 0.40 0.83) (V2 0.35 0.80))
       , (CubicBezier (V2 0.35 0.80) (V2 0.29 0.76) (V2 0.19 0.72) (V2 0.14 0.69))
       , (CubicBezier (V2 0.14 0.69) (V2 0.09 0.65) (V2 (-0.03) 0.57) (V2 (-0.05) 0.28))
       , (CubicBezier (V2 (-0.05) 0.28) (V2 (-0.04) 0.18) (V2 (-0.02) 0.05) (V2 0.00 0.00))
    
       , (CubicBezier (V2 0.10 0.15) (V2 0.14 0.18) (V2 0.18 0.22) (V2 0.18 0.25))
       , (CubicBezier (V2 0.18 0.25) (V2 0.16 0.26) (V2 0.14 0.27) (V2 0.12 0.27))
       , (CubicBezier (V2 0.12 0.27) (V2 0.11 0.23) (V2 0.11 0.19) (V2 0.10 0.15))
    
       , (CubicBezier (V2 0.05 0.18) (V2 0.10 0.20) (V2 0.08 0.26) (V2 0.09 0.30))
       , (CubicBezier (V2 0.09 0.30) (V2 0.07 0.32) (V2 0.06 0.34) (V2 0.04 0.33))
       , (CubicBezier (V2 0.04 0.33) (V2 0.04 0.27) (V2 0.04 0.19) (V2 0.05 0.18))

       , (CubicBezier (V2 0.11 0.30) (V2 0.16 0.44) (V2 0.24 0.61) (V2 0.30 0.66))
       , (CubicBezier (V2 0.30 0.66) (V2 0.41 0.78) (V2 0.62 0.84) (V2 0.80 0.92))

       , (CubicBezier (V2 0.23 0.20) (V2 0.35 0.20) (V2 0.44 0.22) (V2 0.50 0.25))
       , (CubicBezier (V2 0.50 0.25) (V2 0.50 0.33) (V2 0.50 0.41) (V2 0.50 0.49))
       , (CubicBezier (V2 0.50 0.49) (V2 0.46 0.53) (V2 0.42 0.57) (V2 0.38 0.61))

       , (CubicBezier (V2 0.29 0.29) (V2 0.36 0.26) (V2 0.43 0.27) (V2 0.48 0.31))

       , (CubicBezier (V2 0.34 0.39) (V2 0.38 0.34) (V2 0.44 0.36) (V2 0.48 0.37))

       , (CubicBezier (V2 0.34 0.49) (V2 0.38 0.44) (V2 0.41 0.42) (V2 0.48 0.43))

       , (CubicBezier (V2 0.45 0.58) (V2 0.46 0.60) (V2 0.47 0.61) (V2 0.48 0.61))

       , (CubicBezier (V2 0.42 0.61) (V2 0.43 0.64) (V2 0.46 0.68) (V2 0.48 0.67))

       , (CubicBezier (V2 0.25 0.74) (V2 0.17 0.83) (V2 0.08 0.91) (V2 0.00 0.99))
       , (CubicBezier (V2 0.00 0.99) (V2 (-0.08) 0.91) (V2 (-0.17) 0.82) (V2 (-0.25) 0.74))
       , (CubicBezier (V2 (-0.25) 0.74) (V2 (-0.20) 0.63) (V2 (-0.11) 0.53) (V2 (-0.03) 0.43))

       , (CubicBezier (V2 (-0.17) 0.74) (V2 (-0.13) 0.66) (V2 (-0.08) 0.60) (V2 (-0.01) 0.56))

       , (CubicBezier (V2 (-0.12) 0.79) (V2 (-0.07) 0.71) (V2 (-0.02) 0.66) (V2 0.05 0.60))

       , (CubicBezier (V2 (-0.06) 0.86) (V2 (-0.03) 0.77) (V2 0.03 0.72) (V2 0.10 0.66))

       , (CubicBezier (V2 (-0.02) 0.92) (V2 0.02 0.84) (V2 0.09 0.77) (V2 0.16 0.70))
       ]


petal = [ toPrim $ CubicBezier (V2 0.5 0.2) (V2 0.9 0.6)
                            (V2  0.05 1.0) (V2 0.5 1.40)
     , toPrim $ Line (V2 0.5 1.40) (V2 1.20 0.8)
     , toPrim $ Line (V2 1.20 0.8) (V2 0.5 0.2) ]


semicircle = [Path (V2 0.0 0.0) True [PathCubicBezierCurveTo (V2 0.0 0.0) (V2 1.0 0.0) (V2 1.0 1.0),
            PathLineTo (V2 0.0 1.0), PathLineTo (V2 0.0 0.0)]]

square = [rectangle (V2 0.0 0.0) 1.0 1.0]

arc :: [CubicBezier]
arc = [CubicBezier (V2 0.5 0.0) (V2 0.5 0.2759575) (V2 0.72404248 0.5) (V2 1.0 0.5)]


src/ArcsModel.hs
module ArcsModel where

import Lib
import Draw
import CreateDrawings
import Shapes
import Codec.Picture( PixelRGBA8( .. ), writePng, Image)
import Graphics.Rasterific
import Graphics.Rasterific.Texture

----- Usable funcs

genArcs :: FilePath -> Integer -> Int -> IO ()
genArcs path n seed = saveDrawing path scaleSimpleFig white (simpleFig n seed colorsD arc)

genArcsNew :: (Geometry geom, Transformable geom) => [geom] -> FilePath -> Integer -> Int -> [PixelRGBA8] -> IO ()
genArcsNew img path n seed colors = saveDrawing path scaleSimpleFig white (simpleFig n seed colors img)

createNewArc :: Float -> Float -> Float -> Float -> [CubicBezier]
createNewArc x1 y1 x2 y2 = [CubicBezier (V2 0.5 0.0) (V2 x1 y1) (V2 x2 y2) (V2 1.0 0.5)]

---- Helpers

blocksDrawing :: Integer -> [PixelRGBA8] -> Int -> [Drawing PixelRGBA8 ()]
blocksDrawing n colors seed = applyFuncs (zip (colorBlocks n colors seed) (blocks n))

scaleSimpleFig :: Int
scaleSimpleFig = 1200

simpleFig :: (Geometry geom, Transformable geom) => Integer -> Int -> [PixelRGBA8] -> [geom] -> [Drawing PixelRGBA8 ()]
simpleFig n seed colors img = [withTexture (uniformTexture black) $ do mconcat $ blocksDrawing n colors seed,
                           withTexture (uniformTexture black) $ do mconcat $ drawLines scaleSimpleFig (35/ (2 ** fromInteger n)) (arcLimit n img seed)]

src/CreateDrawings.hs
module CreateDrawings where


import Lib
import Prelude hiding (flip, cycle)
import Graphics.Rasterific
import Graphics.Rasterific.Texture
import Codec.Picture( PixelRGBA8( .. ))
import System.Random as SR

white :: PixelRGBA8
white = PixelRGBA8 255 255 255 255

black :: PixelRGBA8
black = PixelRGBA8 0 0 0 255
blue :: PixelRGBA8
blue  = PixelRGBA8 0 0 255 255
fordwine  = PixelRGBA8 94 18 36 255

colorsD :: [PixelRGBA8]
colorsD = [PixelRGBA8 0xE9 0xE3 0xCE 255, PixelRGBA8 0xFF 0x53 0x73 255, 
          PixelRGBA8 0xEE 0xAD 0x2D 255, PixelRGBA8 0x41 0x69 0xE1 255]

----------------------------------------

img2 :: Transformable a => a -> a
img2 = flip.rot45


img3 :: Transformable a => a -> a
img3 = rot.rot.rot.img2

rot3 :: Transformable a => a -> a
rot3 = rot.rot.rot

quartet :: Transformable a => [a] -> [a] -> [a] -> [a] -> [a]
quartet a b c d = above (beside a b) (beside c d)

u :: Transformable a => [a] -> [a]
u i = over (over image2 (rot image2)) (over (rot $ rot image2) (rot $ rot $ rot image2))
       where image2 = img2 i


t :: Transformable a => [a] -> [a]
t i = over i (over (img2 i) (img3 i))



side :: Transformable a => Integer -> [a] -> [a]
side 0 _ = blank
side n i = quartet (side (n-1) i) (side (n-1) i) (rot (t i)) (t i)


corner :: Transformable a => Integer -> [a] -> [a]
corner 0 _ = blank
corner n i = quartet (corner (n-1) i) (side (n-1) i) (rot $ side (n-1) i) (u i)


nonet :: Transformable a => [a] -> [a] -> [a] -> 
                            [a] -> [a] -> [a] -> 
                            [a] -> [a] -> [a] -> [a]
nonet  p q r
       s t u
       v w x =
              aboveScaled 1 2 (besideScaled 1 2 p (beside q r)) 
              (above (besideScaled 1 2 s (beside t u)) 
              (besideScaled 1 2 v (beside w x)))

squarelimit :: Transformable a => Integer -> [a] -> [a]
squarelimit n i = nonet (corner n i) (side n i)             (rot $ rot $ rot $ corner n i)
                  (rot $ side n i)   (u i)                  (rot $ rot $ rot $ side n i)
                  (rot $ corner n i) (rot $ rot $ side n i) (rot $ rot $ corner n i)


cycle :: Transformable a => [a] -> [a]
cycle i = quartet (rot $ rot $ flip i) (rot $ flip i) (rot $ rot $ rot $ flip i) (flip i)

u_centered :: Transformable a => [a] -> [a]
u_centered i = quartet (flip i) (rot $ rot $ rot $ flip i) (rot $ flip i) (rot $ rot $ flip i)

v :: Transformable a => [a] -> [a]
v i = cycle $ rot $ t i

---------------------------------------------

arcs :: (Transformable a) => Integer -> [a] -> Int -> Int -> [a]
arcs 0 i p seed = over (rot3 (img (p+10)) ) (over ((rot.rot) (img (p+11))) (over (img (p+12)) (rot (img (p+13)))))
    where img n
           | chance == 0 = []
           | otherwise = i
             where chance = randomArcs seed !! n
arcs n i p seed = quartet (arcs (n-1) i (p+1+fromInteger n) seed) (arcs (n-1) i (p+2+fromInteger n) seed)
                     (arcs (n-1) i (p+3+fromInteger n) seed) (arcs (n-1) i (p+1+fromInteger n) seed)

sideArc :: Transformable a => Integer -> [a] -> Int -> Int -> [a]
sideArc 0 i p seed = quartet (over (img (p+6)) (rot3 (img (p+9)))) (arcs 0 i (p+1) seed) 
                        (over (img (p+7)) (rot3 (img (p+8)))) (arcs 0 i (p-1) seed)
    where img n
            | chance == 0 = []
            | otherwise = i
                where chance = randomArcs seed !! n
sideArc n i p seed= quartet (sideArc (n-1) i (p+3+fromInteger n) seed) (arcs n i (p+2+fromInteger n) seed) 
                        (sideArc (n-1) i (p-1+fromInteger n) seed) (arcs n i (p+fromInteger n) seed)

cornerArc :: Transformable a => Integer -> [a] -> Int -> Int -> [a]
cornerArc 0 i p seed = quartet (over (img (p+9)) (rot3 (img (p+8)))) (arcs 0 i (p+5) seed) 
                          (img (p+4)) (rot $ over (img (p+6)) (rot3 (img (p+4))))
    where img n
            | chance == 0 = []
            | otherwise = i
                where chance = randomArcs seed !! n
cornerArc n i p seed = quartet (sideArc (n-1) i (p+3+fromInteger n) seed) (arcs n i (p+7+fromInteger n) seed) 
                          (cornerArc (n-1) i (p+6+fromInteger n) seed) (rot $ sideArc (n-1) i (p+3+fromInteger n) seed)

arcLimit :: Transformable a => Integer -> [a] -> Int -> [a]
arcLimit n i seed = nonet (rot3 (cornerArc n i 0 seed)) (rot3 (sideArc n i 1 seed)) ((rot.rot) (cornerArc n i 2 seed)) 
                     (sideArc n i 3 seed) (arcs (n+1) i 4 seed) ((rot.rot) (sideArc n i 5 seed)) 
                     (cornerArc n i 6 seed) (rot (sideArc n i 7 seed)) (rot (cornerArc n i 8 seed))


randomNumbers :: Int -> Int -> [Int]
randomNumbers n seed = randomRs (0,n) (mkStdGen seed)


randomArcs :: Int -> [Int]
randomArcs = randomNumbers 2

applyFuncs = map (\x -> (fst x) (snd x))

---------------------------------------------

blocks n = mesh number size
    where number = round ((6*2**fromInteger n)**2)
          size = 1200 / (6*2**fromInteger n)

mesh 0 _ = []
mesh n size = rectangle (V2 (fromInteger x*size) (fromInteger y*size)) size size : mesh (n-1) size
        where x = mod n (round (1200/size))
              y = round ((1200/size) - 1) - div (n-1) (round (1200/size))

colorBlocks :: Geometry geom => Integer -> [PixelRGBA8] -> Int -> [geom -> Drawing PixelRGBA8 ()]
colorBlocks n = coloring number
    where number = round ((6*2**(fromInteger n+1))**2)

coloring :: Geometry geom => Int -> [PixelRGBA8] -> Int -> [geom -> Drawing PixelRGBA8 ()]
coloring 0 _ _ = []
coloring n xs seed = withTexture (uniformTexture (xs !! (randomNumbers (length xs - 1) seed !! n))) . fill : coloring (n-1) xs seed


-------------------------------------------
makeQuartets :: Transformable a => Integer -> [Int] -> [a] -> [a]
makeQuartets 0 randoml i = quartet (scale 0.85 $ rotn n1 i) (scale 0.85 $ rotn n2 i) (scale 0.85 $ rotn n3 i) (scale 0.85 $ rotn n4 i) where 
  n1 = toInteger $ randoml !! 0
  n2 = toInteger $ randoml !! 1
  n3 = toInteger $ randoml !! 2
  n4 = toInteger $ randoml !! 3
makeQuartets n randoml i = quartet (makeQuartets (n-1) l1 i) (makeQuartets (n-1) l2 i) (makeQuartets (n-1) l3 i) (makeQuartets (n-1) l4 i) where 
  l1 = drop 4 randoml
  l2 = drop 4 l1
  l3 = drop 4 l2
  l4 = drop 4 l3